mmseg/models/losses/cross_entropy_loss.py

```diff
@@ -83,12 +116,22 @@ def binary_cross_entropy(pred,
                 pred.dim() == 4 and label.dim() == 3), \
             'Only pred shape [N, C], label shape [N] or pred shape [N, C, ' \
             'H, W], label shape [N, H, W] are supported'
+        # `weight` returned from `_expand_onehot_labels`
+        # has been treated for valid (non-ignore) pixels
         label, weight = _expand_onehot_labels(label, weight, pred.shape,
                                               ignore_index)
+    else:
+        # should mask out the ignored elements
+        valid_mask = ((label >= 0) & (label != ignore_index)).float()
+        if weight is not None:
+            weight *= valid_mask
+        else:
+            weight = valid_mask
+    # average loss over non-ignored elements
+    if avg_factor is None:
+        avg_factor = label.numel() - ((label >= 0) &
```

Comment:
Incorrect. `label` has been converted to one-hot here.

Reply:
I just checked, when 
```python
    fake_pred = torch.full(size=(2, 21, 8, 8), fill_value=0.5)
    fake_label = torch.ones(2, 8, 8).long()
```
The label is still `(2, 21, 8, 8)`.
![image](https://user-images.githubusercontent.com/24582831/159914133-70f1b3dc-6bab-4814-bec9-87340d917e28.png)

So what modification shoule be taken?

Reply:
label.numel() /label.size(1)

Reply:
You can just move this line before the _expand_onehot_labels function.

Reply:
PyTorch bce would do average over class, so if we follow `avg_factor = (label.numel() - ((label >= 0) & (label == ignore_index)).sum().item())/ label.size(1)`, then in the case:

<img width="293" alt="企业微信截图_16481310574230" src="https://user-images.githubusercontent.com/24582831/159940697-a7c5a26d-011f-43c8-8fda-2ee3b64af179.png">

The `avg_factor` would change and loss would become `200.0`
<img width="625" alt="企业微信截图_16481323572143" src="https://user-images.githubusercontent.com/24582831/159940619-db11197a-ada3-445d-beba-c79c83e615a1.png">

where pytorch bce would average 2 to `100.0`.

<img width="667" alt="企业微信截图_16481323848091" src="https://user-images.githubusercontent.com/24582831/159940981-d2c65e94-d2be-402d-9fab-8b25ee122175.png">

Reply:
In this function, the BCE is used to calculate the multi-label classification loss between the ground truth label and the probability generated by sigmoid logits. 
It is different from the BCE in single-label classification, you should average the loss with the length of the ground truth label instead of the one-hot encoded label.

Reply:
The problem is that the gt-label has been converted to one-hot here. You can calculate the `avg_factor` with the `valid_mask` in `_expand_onehot_labels`